internal/registrycache/secretsyncer.go

```diff
@@ -0,0 +1,148 @@
+	return SecretSyncer{
+		GardenClient:    gardenClient,
+		RuntimeClient:   runtimeClient,
+		RuntimeID:       runtimeID,
+		GardenNamespace: gardenNamespace,
+	}
+}
+
+func (s SecretSyncer) CreateOrUpdate(registryCaches []imv1.ImageRegistryCache) error {
+
+	cachesWithSecret := getRegistryCachesWithSecret(registryCaches)
+
+	for _, cache := range cachesWithSecret {
+		var gardenerSecret v12.Secret
+		err := s.GardenClient.Get(context.TODO(), client.ObjectKey{Name: getGardenSecretName(cache), Namespace: s.GardenNamespace}, &gardenerSecret)
+
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+
+		if err == nil {
+			err = s.updateSecretInGardenCluster(cache, gardenerSecret)
+			if err != nil {
+				return fmt.Errorf("failed to update secret for registry cache %s: %w", cache.Name, err)
+			}
+		} else {
+			err = s.copySecretFromRuntimeToGardenCluster(cache)
+			if err != nil {
+				return fmt.Errorf("failed to copy secret for registry cache %s: %w", cache.Name, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s SecretSyncer) copySecretFromRuntimeToGardenCluster(cacheConfig imv1.ImageRegistryCache) error {
+
+	var secret v12.Secret
+	err := s.RuntimeClient.Get(context.TODO(), client.ObjectKey{Name: *cacheConfig.Config.SecretReferenceName, Namespace: cacheConfig.Namespace}, &secret)
+
+	if err != nil {
+		return err
+	}
+
+	newSecret := v12.Secret{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      getGardenSecretName(cacheConfig),
+			Namespace: s.GardenNamespace,
+			Labels: map[string]string{
+				RuntimeSecretLabel: s.RuntimeID,
+			},
+			Annotations: map[string]string{
+				CacheIDAnnotation: cacheConfig.UID,
+			},
+		},
+		Data: secret.Data,
+	}
+
+	return s.GardenClient.Create(context.TODO(), &newSecret)
+}
+
+func (s SecretSyncer) updateSecretInGardenCluster(cacheConfig imv1.ImageRegistryCache, gardenerSecret v12.Secret) error {
+	var runtimeSecret v12.Secret
+	err := s.RuntimeClient.Get(context.TODO(), client.ObjectKey{Name: *cacheConfig.Config.SecretReferenceName, Namespace: cacheConfig.Namespace}, &runtimeSecret)
+
+	if err != nil {
+		return err
+	}
+
+	gardenerSecret.Data = runtimeSecret.Data
+
+	return s.GardenClient.Update(context.TODO(), &gardenerSecret)
+}
+
+func (s SecretSyncer) Delete(registryCaches []imv1.ImageRegistryCache) error {
+
+	cachesWithSecret := getRegistryCachesWithSecret(registryCaches)
+
+	var gardenSecrets v12.SecretList
+	err := s.GardenClient.List(context.TODO(), &gardenSecrets, client.MatchingLabels{RuntimeSecretLabel: s.RuntimeID})
+
+	if err != nil {
+		return fmt.Errorf("failed to list garden secrets: %w", err)
+	}
+
+	for _, gardenSecret := range gardenSecrets.Items {
+		registryCacheUID := gardenSecret.Annotations[CacheIDAnnotation]
+
+		if registryCacheUID != "" && !registryCacheUidExists(registryCacheUID, cachesWithSecret) {
+			err = s.GardenClient.Delete(context.TODO(), &gardenSecret)
+
+			if err != nil {
```

Comment:
We should filter out `NotFound` errors.

Reply:
This is a corner case, but we agreed it will be addressed in the next PR.